### PR TITLE
Identity Server for Sitecore 9.3.0 in Nano Server Container

### PR DESCRIPTION
**Pull request by anarsen**
# Description

As proposed in #221 this PR demonstrates how Identity Server 4.0.0 for Sitecore 9.3.0 can be hosted within a Nano Server container. See the issue for pros and cons.

Build it with:

    SitecoreImageBuilder\Invoke-Build `
      -Path .\windows\ `
      -Tags 'sitecore-identity:9.3.0-nanoserver-1809' `
      -InstallSourcePath UP_TO_YOU

**Note** I'm not very well-versed in the different Sitecore topologies, but from what I gather Sitecore Identity Server is a _service_ in the Sitecore landscape, which can be used across all platforms (XM, XP, XC, ...). If I've misplaced it in the folder structure, then I'll be happy to move it into its appropriate place.

# Configuration

Being an ASP.NET Core application at the bottom, almost all of (if not all) [Identity Server can be configured through environment variables](https://doc.sitecore.com/developers/93/sitecore-experience-manager/en/configuration.html).

The IIS handled the HTTPS termination originally, and if you still want end-to-end HTTPS, you can [configure the Kestrel webserver to listen on HTTPS](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/servers/kestrel?view=aspnetcore-2.1#endpoint-configuration-2). It listens only on HTTP by default.

    SITECORE_URLS=https://+:443
    SITECORE_Kestrel__Certificates__Default__Path=c:\certificates\identity.pfx
    SITECORE_Kestrel__Certificates__Default__Password=SNIP

### CODE REVIEW

File: windows/9.3.x/sitecore-identity/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+FROM ${CERTIFICATES_IMAGE} as certificates
+FROM ${BUILD_IMAGE} as build
+
+SHELL ["powershell", "-Command", "$ErrorActionPreference = 'Stop'; $ProgressPreference = 'SilentlyContinue';"]
+
+ARG ASSETS_USE_WDP
+
+COPY --from=assets ["${ASSETS_USE_WDP}", "c:\\temp\\packages\\"]
+
+RUN Expand-Archive -Path 'c:\\temp\\packages\\*.zip' -DestinationPath 'c:\\temp'
```
Comment (joostmeijles) [inline windows/9.3.x/sitecore-identity/Dockerfile]: I think you need to add config here XDT & license, see https://github.com/Sitecore/docker-images/blob/master/windows/9.3.x/sitecore-xp-identity/Dockerfile

File: windows/dependencies/sitecore-certificates/Dockerfile [hunk 2]
```diff
@@ -33,7 +33,7 @@ RUN $env:ROOT_CERT_NAME = 'sitecore-root'; `
     (Get-ChildItem -Path $env:CERT_STORE | Where-Object { $_.Subject -eq ('CN={0}' -f $env:IDENTITY_CLIENT_CERT_NAME ) }).Thumbprint | Out-File -FilePath ('C:\\certificates\\{0}.thumbprint' -f $env:IDENTITY_CLIENT_CERT_NAME) -NoNewline -Encoding utf8; `
     (Get-ChildItem -Path $env:CERT_STORE | Where-Object { $_.Subject -eq ('CN={0}' -f $env:IDENTITY_SSL_CERT_NAME ) }).Thumbprint | Out-File -FilePath ('C:\\certificates\\{0}.thumbprint' -f $env:IDENTITY_SSL_CERT_NAME) -NoNewline -Encoding utf8; `
     (Get-ChildItem -Path $env:CERT_STORE | Where-Object { $_.Subject -eq ('CN={0}' -f $env:CM_SSL_CERT_NAME ) }).Thumbprint | Out-File -FilePath ('C:\\certificates\\{0}.thumbprint' -f $env:CM_SSL_CERT_NAME) -NoNewline -Encoding utf8; `
-    $env:CERT_PASSWORD | Out-File -FilePath 'C:\\certificates\\password' -NoNewline -Encoding utf8;
+    $env:CERT_PASSWORD | Out-File -FilePath 'C:\\certificates\\password' -NoNewline -Encoding Ascii;
```
Comment (joostmeijles) [inline windows/dependencies/sitecore-certificates/Dockerfile]: Unrelated change?

File: windows/9.3.x/sitecore-identity/build.json [hunk 3]
```diff
@@ -0,0 +1,15 @@
+{
```
Comment (joostmeijles) [inline windows/9.3.x/sitecore-identity/build.json]: Current image is named `sitecore-xp-identity`. Please replace that.

File: windows/dependencies/sitecore-certificates/Dockerfile (hunk 2 above)
Comment (anarsen) [reply windows/dependencies/sitecore-certificates/Dockerfile]: Unfortunately not. See the commit message in 18bfc322bf1d4e1f29ba4adec779703337ab43f0.

Comment (anarsen) [reply windows/dependencies/sitecore-certificates/Dockerfile]: I noticed a point is missing in that commit message. I mention `cmd.exe` in the commit message but not why. It's because Nano Server only ships with the `cmd.exe` shell, and not Powershell.

File: windows/9.3.x/sitecore-identity/Dockerfile (hunk 1 above)
Comment (anarsen) [reply windows/9.3.x/sitecore-identity/Dockerfile]: Actually, for Identity Server this can be configured entirely through environment variables thanks to [Sitecore Host](https://doc.sitecore.com/developers/93/sitecore-experience-manager/en/configuration.html#UUID-5c6cf510-2546-8c91-4393-bfa6b49a6f62_section-idm46212281838784).

So in order to set everything otherwise set by [Sitecore.IdentityServer.Host.xml.xdt](https://github.com/Sitecore/docker-images/blob/master/windows/9.3.x/sitecore-xp-identity/config/production/Sitecore.IdentityServer.Host.xml.xdt) with environment variables, you'd do

```
SITECORE_Sitecore__IdentityServer__CertificateThumbprint=<value>
SITECORE_Sitecore__IdentityServer__CertificateStoreLocation=LocalMachine
SITECORE_Sitecore__IdentityServer__CertificateStoreName=My
SITECORE_Sitecore__IdentityServer__SitecoreMembershipOptions__ConnectionString=<value>
SITECORE_Sitecore__IdentityServer__AccountOptions__PasswordRecoveryUrl=<value>
SITECORE_Sitecore__IdentityServer__Clients__DefaultClient__AllowedCorsOrigins=<value>
SITECORE_Sitecore__IdentityServer__Clients__PasswordClient__ClientSecrets__ClientSecrets=<value>
```

However, since Nano Server doesn't have the certificate store provided by Windows Server (Core), the three top lines will be pointing to a file instead:

So **instead** of these 

```
SITECORE_Sitecore__IdentityServer__CertificateThumbprint=<value>
SITECORE_Sitecore__IdentityServer__CertificateStoreLocation=LocalMachine
SITECORE_Sitecore__IdentityServer__CertificateStoreName=My
```

we'll do this intead [sitecore-identity/Dockerfile, line 35](https://github.com/anarsen/docker-images/blob/identity-server-4.0.0-nanoserver/windows/9.3.x/sitecore-identity/Dockerfile#L35).

How does this sit with you?

Comment (joostmeijles) [reply windows/9.3.x/sitecore-identity/Dockerfile]: Cool, didn't know that!

The environment variables settings should be added to the example docker-compose files than.

Comment (anarsen) [reply windows/9.3.x/sitecore-identity/Dockerfile]: Of course, I'll get right on it.

**Comment by anarsen** [comment]
> Please add also `sitecore-xc-identity` (for 9.3.x). There current version has a small Dockerfile so it shouldn't be too much work.

Should I add a `sitecore-xm-identity` as well?

File: windows/9.3.x/sitecore-identity/build.json (hunk 3 above)
Comment (anarsen) [reply windows/9.3.x/sitecore-identity/build.json]: Done in ba90585af6836ef44fa6d354d31337fc6d256a6a

File: windows/9.3.x/sitecore-xm/config/IdentityServer.config [hunk 4]
```diff
@@ -1,10 +1,10 @@
 <?xml version="1.0" encoding="utf-8"?>
 <configuration xmlns:patch="http://www.sitecore.net/xmlconfig/">
     <sitecore>
-        <sc.variable name="identityServerAuthority">
-            <patch:attribute name="value">http://id</patch:attribute>
-        </sc.variable>
         <settings>
+            <setting name="FederatedAuthentication.IdentityServer.Authority">
```
Comment (anarsen) [inline windows/9.3.x/sitecore-xm/config/IdentityServer.config]: @joostmeijles Notice this change. Is there a consequence of changing this from `<sc.variable />` to a `<setting />` element? It works and the setting is available in `showconfig.aspx`, but is there something I'm missing by changing the element type?

Anyway, the reason for this change is to allow for `FederatedAuthentication.IdentityServer.Authority` through environment variables. It's new to me, but apparantly Sitecore replaces any occurences of `$(env:FOO)` with the value from `FOO` environment variable - if it exists.

I figured that'd be a good way forward instead of relying different patch configs. Your feedback on this is most welcome :-)

**Comment by anarsen** [comment]
@joostmeijles I'll update all occurrences of Identity Server services in Docker Compose files to use the Nano Server version, correct? I might be missing a few places.

**Comment by joostmeijles** [comment]
> @joostmeijles I'll update all occurrences of Identity Server services in Docker Compose files to use the Nano Server version, correct? I might be missing a few places.

Sounds good, I will verify it once its in the PR.

**Comment by bplasmeijer** [comment]
any update @anarsen / @joostmeijles ?

**Comment by anarsen** [comment]
@bplasmeijer Yeah, pending me actually. Just been busy with my client past weeks.

It's not pending anything major, it's actually just a `docker-compose.yml` file showing how it works. I'll try and have it ready for review the coming Monday.

**Comment by sshushliapin** [comment]
@anarsen, any progress on the docker-compose so far? Should it be smth similar to 9.2 version, e.g.:

```
  identity:
    image: ${REGISTRY}sitecore-xp-identity:${SITECORE_VERSION}-nanoserver-${NANOSERVER_VERSION}
    mem_limit: 1GB
    volumes:
      - ${LICENSE_PATH}:C:\license
      - .\data\identity:C:\inetpub\wwwroot\logs
    ports:
      - "44005:80"
    depends_on:
      - sql
```

**Comment by bplasmeijer** [comment]
Hi
If not valid anymore, please close PR, otherwise, rebase and ask for new QA.
Thanks
Bart

**Comment by anarsen** [comment]
@bplasmeijer Thanks for pinging. Together with my client we're working in a different direction at this time. I still think the PR has value, but I'd appreciate if someone could take over on it. Perhaps from your organization if you feel the same.

I'll be happy to rebase it so it's not going completely stale.

**Comment by anarsen** [comment]
My client's priorities shifted and this is no longer relevant enough for me to work on.

If anyone wants to pick this up again after I close it, feel free to get in touch.
